cedar-policy-validator/src/human_schema/ast.rs

```diff
@@ -0,0 +1,441 @@
+    pub fn is_in_cedar(&self) -> bool {
+        self.0.node.is_in_cedar()
+    }
+
+    /// Is this name exactly the cedar namespace?
+    pub fn is_cedar(&self) -> bool {
+        self.0.node.is_cedar()
+    }
+}
+
+impl std::fmt::Display for Path {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0.node)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+struct PathInternal {
+    basename: Id,
+    namespace: Vec<Id>,
+}
+
+impl PathInternal {
+    fn iter(&self) -> impl Iterator<Item = &Id> {
+        self.namespace.iter().chain(once(&self.basename))
+    }
+
+    fn into_iter(self) -> impl Iterator<Item = Id> {
+        self.namespace.into_iter().chain(once(self.basename))
+    }
+
+    /// Is this referring to a name _in__ the __cedar namespace: ex: __cedar::Bool
+    fn is_in_cedar(&self) -> bool {
+        // `0` is the position of the most significant namespace
+        self.namespace
+            .get(0)
+            .map(|id| id.as_ref() == CEDAR_NAMESPACE)
+            .unwrap_or(false)
+    }
+
+    /// Is this name exactly the cedar namespace?
+    fn is_cedar(&self) -> bool {
+        self.namespace.is_empty() && self.basename.as_ref() == CEDAR_NAMESPACE
+    }
+
+    /// Is this referring to a name _in__ the __cedar namespace: ex: __cedar::Bool or the unqualified namespace
+    fn is_in_unqualified_or_cedar(&self) -> bool {
+        self.namespace.is_empty() || self.is_in_cedar()
+    }
+}
+
+impl std::fmt::Display for PathInternal {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.namespace.is_empty() {
+            write!(f, "{}", self.basename)
+        } else {
+            let namespace = self.namespace.iter().map(|id| id.as_ref()).join("::");
+            write!(f, "{namespace}::{}", self.basename)
+        }
+    }
+}
+
+/// This struct represents Entity Uids in the Schema Syntax
+#[derive(Debug, Clone)]
+pub struct QualName {
+    pub path: Option<Path>,
+    pub eid: SmolStr,
+}
+
+impl QualName {
+    pub fn unqualified(eid: SmolStr) -> Self {
+        Self { path: None, eid }
+    }
+
+    pub fn qualified(path: Path, eid: SmolStr) -> Self {
+        Self {
+            path: Some(path),
+            eid,
+        }
+    }
+}
+
+/// A [`Namespace`] has a name and a collection declaration
+/// A schema is made up of a series of fragments
+/// A fragment is a series of namespaces
+#[derive(Debug, Clone)]
+pub struct Namespace {
+    /// The name of this namespace. If [`None`], then this is the unqualified namespace
+    pub name: Option<Node<Path>>,
+    /// The [`Declaration`]s contained in this namespace
+    pub decls: Vec<Node<Declaration>>,
+}
+
+impl Namespace {
+    /// Is this [`Namespace`] unqualfiied?
+    pub fn is_unqualified(&self) -> bool {
+        self.name.is_none()
+    }
+
+    /// Does this [`Namespace`] start with `__`, the reserved pattern
+    pub fn is_reserved_namespace(&self) -> bool {
```

Comment:
[RFC 52](https://github.com/cedar-policy/rfcs/pull/52) isn't accepted yet, are we still ok here?

Reply:
I can disable this for now yeah